### ISSUE TITLE
Create CUE specification for testdata.yaml and problem.yaml

### ISSUE THREAD

**Issue by thorehusfeldt** [Kattis/problem-package-format#30]
Here are some ideas:

```cue
// Defines valid contents of testdata.yaml
package problemformat

import "strings"
import "strconv"

#filename: =~ "^[a-zA-Z0-9][a-zA-Z0-9_.-]*[a-zA-Z0-9]$"
#path: =~ "[a-zA-Z0-9_.-/]*"

#testdata_settings: {
	on_reject: *"break" | "continue"
	grading: *"default" | "custom"
	grader_flags:  *"" | string
    if grading == "default" { grader_flags? : #default_grader_flags }
	input_alidator_flags: *"" | string
	output_validator_flags: *"" |string
	accept_score: *"1" | #score
	reject_score: *"0" | #score
	range: *"-inf +inf" | string
    // Verify that the scores and ranges make sense:
    _range_syntax: strings.Split(range, " ") & [#score, #score] // space-separated scores  
    _lo: strconv.ParseFloat(strings.Split(range, " ")[0], 64)   // parses to float (including '-inf')
    _hi: strconv.ParseFloat(strings.Split(range, " ")[1], 64)
    _wa: strconv.ParseFloat(reject_score, 64)
    _ac: strconv.ParseFloat(accept_score, 64)
    _order: true & _lo <= _wa  & _wa <= _ac & _ac <= _hi
}
// matches "1", "06", "21", ".4", "-1.2", but not 'inf'
#score: =~ "^-?([0-9]+|[0-9]*.[0-9]+)$" 

// Default grader
// --------------

#default_grader_flags: this={
    string  
    _as_struct: { for w in strings.Split(this, " ")  { (w): _ } }  // convert to struct and ...
    _as_struct: #valid_default_grader_fields                       // ... validate its fields
    }

// Default grader flags (converted to fields of a CUE struct for validation)
#valid_default_grader_fields: {
    #verdict_aggregation_mode?   // at most one verdict aggregation mode
    #score_aggregation_mode?     // at most one score aggregation mode
    ignore_sample?: _            // two more optional flags
    accept_if_any_accepted?: _
}

#verdict_aggregation_mode: {first_error: _ } | *{worst_error: _  } | {always_accept: _ }
#score_aggregation_mode:  {min: _ } | {max: _ } | *{sum: _ } | {avg: _ }
```

and here’s a specification of `problem.yaml`

```cue
#problem_settings: {
    // Should these fields all also accept null? If so, what's the semantics of, say,
    //     license:
    // TODO: typical-problem.yaml violates spec (name missing)
    // TODO: maximal-problem.yaml violates spec (name missing, validator should be validator_flags)
    name: string | { [string]: string }
    type?: *"pass-fail" | "scoring"
    author?: string
    source?: string
    if source != _|_ { source_url?: string } # only allow source_url if source is specified
    license?: *"unknown" | "public domain" | "cc0" | "cc by" | "cc by-sa" | "educational" | "permission"
    rights_owner?: string
    limits?: #limits
    validation?: *"default" | #custom_validation
    validator_flags?: string
    scoring?: {
        objective: *"max" | "min"
        show_test_data_groups: *false | true
    }
    keywords?: string | [...string]
    uuid?: string
    languages?: string | [...string]
}

#custom_validation: this={
    string
    _as_struct: { for w in strings.Split(this, " ")  { (w): _ } }
    _as_struct: close({ 
        custom: _,       // Must include "custom",
        score?: _,       // can include "score" ...
        interactive?: _  // ... and "interactive"
})
}

#limits: { 
    // All are optional, right?
    // Is the empty limits dictionary valid? (I guess yes)
    // TODO: are these all ints?
    time_multiplier?: *5 | int
    time_safety_margin?: *2 | int
    memory?: int
    output?: int
    code?: int
    compilation_time?: int
    compilation_memory?: int
    validation_time?: int
    validation_memory?: int
    validation_output?: int
}
```

**Comment by thorehusfeldt** [comment]
Related: the two example yaml files at https://github.com/Kattis/problem-package-format/blob/master/examples/problem.yaml.md should validate against this specification.

**Comment by niemela** [comment]
Fixed by #128 (? @thorehusfeldt)
